File: Data Structures & Algorithms/delete-node-in-a-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        
        def least(node):
            if not node:
                return None
            while node.left:
                node = node.left
            return node

        def remove(node, key):
            if not node:
                return None
            if key < node.val:
                node.left = remove(node.left, key)
            elif key > node.val:
                node.right = remove(node.right, key)
            else:
                if not node.left and not node.right:
                    return None
                elif not node.left:
                    return node.right
                elif not node.right:
                    return node.left
                else:
                    print(node.val)
                    temp = least(node.right)
                    node.val = temp.val
                    node.right = remove(node.right, temp.val)

            return node

        return remove(root, key)

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/*
    Same algorithm as before except we only record the last val
*/
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ret = new ArrayList<>();
        if(root == null){
            return ret;
        }
        Queue<TreeNode> que = new LinkedList<>();
        que.add(root);
        
        while(!que.isEmpty()){
            int size = que.size();
            while(size>0){
                TreeNode target = que.remove();
                if(size==1){
                    ret.add(target.val);
                }
                if(target.left!=null){
                    que.add(target.left);
                }
                if(target.right!=null){
                    que.add(target.right);
                }
                size --;
            }
        }
        
        return ret;
    }
}
